src/task.rs

```diff
@@ -227,15 +232,22 @@ fn kernel(_ : *const u32) {
             while task_num < TASKS.len() {
                 let task = &mut TASKS[task_num];
 
-                match task.state {
+                match &task.state {
                     TaskState::Runnable => {},
-                    TaskState::Asleep => {
+                    TaskState::Asleep(wake_up_ticks) => {
                         //If this task is asleep, see if we should wake it up
-                        if task.wake_up_tick < TICKS {
+                        if *wake_up_ticks < TICKS {
                             //If we wake the task up, we can run it
                             task.state = TaskState::Runnable;
                         }
                     },
+                    TaskState::Blocking(sem) => {
+                       //If the semaphore this task is blocking on is available
+                       //to be taken, wake up the task so it can attempt to take it
+                       if sem.is_available() {
```

Comment:
What happens if a semaphore is deleted before the task wakes up? Or is that behavior not supported yet?

Reply:
I guess that's an oversight. I can make it take use an Arc reference counter instead of a reference to avoid this problem

Reply:
Actually, if a semaphore defines as an Arc is blocking a task, it can't be deleted since it's still being used by the task that is blocking in the first place through the task calling take. Is my logic correct here?